import {
	ReactNode,
	createContext,
	useContext,
	useEffect,
	useState,
} from 'react';
import {
	checkAuthStatus,
	loginUser,
	logoutUser,
	signupUser,
} from '../helpers/api-communicator';

type User = {
	name: string;
	email: string;
};
type UserAuth = {
	isLoggedIn: boolean;
	user: User | null;
	login: (email: string, password: string) => Promise<void>;
	signup: (name: string, email: string, password: string) => Promise<void>;
	logout: () => Promise<void>;
};

const AuthContext = createContext<UserAuth | null>(null);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
	const [user, setUser] = useState<User | null>(null);
	const [isLoggedIn, setIsLoggedIn] = useState(false);

	useEffect(() => {
		async function checkStatus() {
			try {
				const data = await checkAuthStatus();
				if (data) {
					setUser({ email: data.email, name: data.name });
					setIsLoggedIn(true);
				}
			} catch (error) {
				console.error('Error checking auth status:', error);
			}
		}
		checkStatus();
	}, []);

	const login = async (email: string, password: string) => {
		try {
			const data = await loginUser(email, password);
			if (data) {
				setUser({ email: data.email, name: data.name });
				setIsLoggedIn(true);
			}
		} catch (error) {
			console.error('Login failed:', error);
			throw error;
		}
	};

	const signup = async (name: string, email: string, password: string) => {
		try {
			const data = await signupUser(name, email, password);
			if (data) {
				setUser({ email: data.email, name: data.name });
				setIsLoggedIn(true);
			}
		} catch (error) {
			console.error('Signup failed:', error);
			throw error;
		}
	};

	const logout = async () => {
		try {
			await logoutUser();
			setIsLoggedIn(false);
			setUser(null);
		} catch (error) {
			console.error('Logout failed:', error);
		}
	};

	const value = {
		user,
		isLoggedIn,
		login,
		logout,
		signup,
	};

	return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export const useAuth = () => {
	const context = useContext(AuthContext);
	if (!context) {
		throw new Error('useAuth must be used within an AuthProvider');
	}
	return context;
};
